Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs

```diff
@@ -708,6 +708,41 @@ internal bool disabledWhileInBackground
         // NOTE: This contains *leaf* controls only.
         internal uint[] m_StateOffsetToControlMap;
 
+        // Holds the nodes that represent the tree of memory ranges that each control occupies. This is used when
+        // determining what controls have changed given a state event or partial state update.
+        internal ControlBitRangeNode[] m_ControlTreeNodes;
+
+        // An indirection table for control bit range nodes to point at zero or more controls. Indices are used to
+        // point into the m_ChildrenForEachControl array.
+        internal ushort[] m_ControlTreeIndices;
+
+        internal struct ControlBitRangeNode
+        {
+            // only store the end bit offset of each range because we always do a full tree traversal so
+            // the start offset is always calculated at each level.
+            public ushort endBitOffset;
+
+            // points to the location in the nodes array where the left child of this node lives, or -1 if there
+            // is no child. The right child is always at the next index.
+            public short leftChildIndex;
+
+            // each node can point at multiple controls (because multiple controls can use the same range in memory and
+            // also because of overlaps in bit ranges). The control indicies for each node are stored contiguously in the
+            // m_ControlTreeIndicies array on the device, which acts as an indirection table, and these two values tell
+            // us where to start for each node and how many controls this node points at. This is an unsigned short so that
+            // we could in theory support devices with up to 65553 controls.
+            public ushort controlStartIndex;
+            public byte controlCount;
```

Comment:
Is there a good reason to keep controlCount as a byte here since it will anyway get padded to the size of an ushort?

Reply:
I was going for the smallest possible size here to keep the entire tree of ranges in L1, but I forgot to add the Pack attribute! Good catch.